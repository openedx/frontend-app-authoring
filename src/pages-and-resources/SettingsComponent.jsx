import React from 'react';
import PropTypes from 'prop-types';
import { useParams, useNavigate } from 'react-router-dom';

const SettingsComponent = ({ url }) => {
  const { appId } = useParams();
  const navigate = useNavigate();

  const LazyLoadedComponent = React.useMemo(() => React.lazy(async () => {
    try {
      // There seems to be a bug in babel-eslint that causes the checker to crash with the following error
      // if we use a template string here:
      //     TypeError: Cannot read property 'range' of null with using template strings here.
      // Ref: https://github.com/babel/babel-eslint/issues/530
      return await import('@openedx-plugins/course-app-' + appId + '/Settings.jsx'); // eslint-disable-line
    } catch (error) {
      console.trace(error); // eslint-disable-line no-console
      return null;
    }
  }), [appId]);

  return <LazyLoadedComponent onClose={() => navigate(url)} />;
};

SettingsComponent.propTypes = {
  url: PropTypes.string.isRequired,
};

export default SettingsComponent;
