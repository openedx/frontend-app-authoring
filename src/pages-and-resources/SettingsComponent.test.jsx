import React, { Suspense } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render } from '@testing-library/react';

import SettingsComponent from './SettingsComponent';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({
    appId: 'wiki',
  }),
}));

describe('SettingsComponent', () => {
  test('renders LazyLoadedComponent when provided with props', async () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Suspense fallback="...">
          <SettingsComponent url="/some-url" />
        </Suspense>
      </BrowserRouter>,
    );

    expect(asFragment).toMatchSnapshot();
  });
});
