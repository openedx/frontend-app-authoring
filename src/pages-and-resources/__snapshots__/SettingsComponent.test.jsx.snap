// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsComponent renders LazyLoadedComponent when provided with props 1`] = `[Function]`;
